### ISSUE TITLE
Vispy Sprint in March 2015

### ISSUE THREAD

**Issue by rossant** [vispy/vispy#685]
This is a list of high-priority things that we could do during the sprint (24-27 March in Paris):

* High-level plotting interface
* Collections
* Axes, ticks, labels, grids visuals
* Fast batch rendering of text (multiple labels)
* High-quality, batchable lines and [paths](http://matplotlib.org/users/path_tutorial.html) visuals (including Bézier curves and polygons)
* More visuals
* Improve mpl_plot backend (need paths, axes, ticks, labels, grids)
* Reorganize/clean up/improve all the examples
* New website and gallery (inspired by glumpy), + possible blog & RSS feed
* [LC/AK] Performance of the scene graph
* [LC/AK] Converge on the scene graph / visuals API.
* (lower priority) Improve the documentation

Feel free to edit this. You can also put your initials in front of what you want to tackle.

**Comment by izaid** [comment]
I'm resurrecting this. For the sprint, my main priority is to introduce GPU support so that VisPy can take GPU data from PyCUDA, DyND, Numba, .... and display it easily. Now, I've lost track a bit of what's going on. Does VisPy still always expect NumPy arrays? I think there was chatter at some point about using some sort of lightweight wrapper that could handle GPU data?

**Comment by almarklein** [comment]
Its interesting, since we've also played with the idea of allowing rendering stuff without numpy at all. I think the idea was to be able to run on Kivi, although this might not be necessary since AFAIK Kivi does support numpy now.

Anyway, some sort of proxy class might make sense... But would require major modifications to much of the code-base.

**Comment by izaid** [comment]
What does everything else think about dropping NumPy? Is that excessive? I think it comes down to a choice between dropping NumPy and using a lightweight proxy of our own for all data (CPU or GPU). Or adopting a second, optional package (like PyCUDA or DyND) that stores GPU data.

VisPy currently keeps a reference (as a NumPy array) to data it is rendering, right? If we want to stick with that strategy, then we need to do the latter.

**Comment by jdreaver** [comment]
Numpy is the defacto package for handling large, homogeneous arrays in
Python. I think dropping numpy entirely wouldn't bring much benefit, since
anyone with data large enough to need GPU visualization is probably using
numpy to store their data in the first place. A proxy class or second
package seems to be the way to go, but like Almar said, it could take a lot
of work.

**Comment by izaid** [comment]
If we're keeping NumPy, then I would screw the proxy class and simply add DyND (https://github.com/libdynd/libdynd) as an optional dependency for the GPU. Of course, I'm biased, as one of the lead developers of DyND.

I don't see the point of maintaining an array implementation when there are so many other packages that do it better. The only reason to main our own would be if we want to have no dependency on them.

**Comment by almarklein** [comment]
What would we have to change in vispy to support both Dynd and Numpy? Can it be done without a proxy class? Also, if we have dynd, would we be able to handle GPU arrays etc?

**Comment by rossant** [comment]
+1 for adding an optional dynd/PyCUDA/etc dependency.

What about having a new gloo object; in GLIR something like:

CREATE 1 DyndArray

This object is ready only in VisPy. We instanciate it by giving a pointer
to the GPU buffer somehow (not sure how that works exactly with
dynd/OpenGL).

I think the first step would be to try a toy example with just dynd and
OpenGL API calls from our OpenGL wrapper.


2015-02-25 10:17 GMT+01:00 Almar Klein <notifications@github.com>:

**Comment by rougier** [comment]
It really depends on the underlying structure of the Dynd array. Since vispy can have GPU-only memory, it's only a matter of having the right dtype (numpy) description of the Dynd array.

**Comment by izaid** [comment]
I just had a little closer look at the relevant parts of VisPy. What I think needs to happen is the following:

We'll need to unify how VisPy interacts with various array backends. For example, we can't call numpy.asarray or numpy.ascontiguousarray in certain places anymore. Likewise, while DyND is good at interoperating with NumPy, it is not 100% the same because it doesn't aim to be. A DyND GPU array obviously can't map to a NumPy array.

So, to handle different array classes, VisPy will need a set of functions to replace any calls to NumPy's functions. If necessary, we can wrap this in a VisPy class, but it would be very lightweight (e.g. you'd probably have to get the original array back to do anything beyond asking for its shape, dtype, or data pointer). I wouldn't want to implement slicing methods or anything like that.

--

As for actually creating GPU data, I think we have to do what @rossant suggested and let VisPy pass a data pointer to DyND. DyND will need to be modified to allow CUDA / OpenGL interoperability.

Bottom line: VisPy needs to learn how to use any array backends it wants to support. DyND needs to get OpenGL support.

--

@almarklein Yes, DyND already has CUDA support (modulo the OpenGL interoperability). Yes, this can be used in DyND's Python wrapper. At the Python level, it is already close to what PyCUDA can do -- except DyND also supports the CPU and many, many other things.

**Comment by rougier** [comment]
@izaid What is the memory layout of a dynd array ?

**Comment by izaid** [comment]
@rougier It supports more then strided arrays, so there isn't a quick answer. For VisPy purposes, consider it strided like NumPy.

**Comment by rougier** [comment]
The main layouts on the GPU (at least for vispy) is strided or blocked. Meaning for example, that for a given data type (let's say Color,Position,Normal), we can use: CPNCPNCPNCPN or CCCC, PPPP, NNNN (or any combination). Any upload to the GPU requires a CPU contiguous block of memory (no CPU stride when uploading) which brings some difficulties if you do not want to have too many small uploads. Apart from that, I think there is no others constraints (but the allowed basic type of course).

You might want to have a look at https://github.com/vispy/vispy/blob/master/vispy/gloo/buffer.py and more specifically the Buffer class that is dtype agnostic. It is only in charge of uploading bytes into GPU when necessary.

I took a different approach in glumpy in order to have a single upload anytime (largest contiguous block) and use the numpy machinery to automatically keep track of what's need to be changed.
See https://github.com/glumpy/glumpy/blob/master/glumpy/gloo/buffer.py


For a more generic interface, it would be good to be able to specify if we want a CPU copy or not, to bind a buffer to existing GPU memory,  to be able to copy GPU to CPU and to automatically track changes.

**Comment by rossant** [comment]
@izaid If I understand correctly, DyND can replace PyCUDA and PyOpenCL and provide its own GPU wrapper? So if we manage to make VisPy and DyND play together, then we'll never need to mess with PyCUDA/PyOpenCL?

**Comment by izaid** [comment]
@rougier All noted, I'll check out the buffer class in more detail. Thanks for all the info! Agree with you on the generic interface.

@rossant Yes, that is effectively true. DyND can replace PyCUDA. As for PyOpenCL, DyND doesn't have OpenCL support (yet). The only reason we would need to mess with PyCUDA / PyOpenCL is if we want to support those independently.

**Comment by rossant** [comment]
@izaid good to know! I've always had tons of problems with making PyCUDA/PyOpenCL work on various platforms; I hope DyND is better :) I'll try to test it before the sprint.

**Comment by izaid** [comment]
@rossant I can't promise everything is going to work easily, but at least you can get me to fix things which are not! ;)

**Comment by Eric89GXL** [comment]
Closing for open issues

**Comment by rossant** [comment]
TODO on Friday (roughly decreasing priority)

* **AXIS/TICKS**: @nippoo and others?
* `vispy.plot` / `vispy.mpl_plot` @Eric89GXL and others?
* Finish and merge #800. This way we can port many Glumpy examples and demos easily. @campagnola 
* Website: @rougier needs to do a PR
* Documentation: @rossant
* Choropleth map example #795: done by @rougier, need to make a PR

**Comment by izaid** [comment]
Need to also merge my emulated texture branch. Either @rougier and I will get the new version working, or I'll merge the more memory-limited one.
